### PR TITLE
TST: clean up `ppmt` tests

### PR DESCRIPTION
**Pull request by person142**
In particular,

- Move them into their own class
- Add tests for invalid `per` arguments

First mentioned in https://github.com/numpy/numpy-financial/pull/22.

### CODE REVIEW

File: numpy_financial/tests/test_financial.py [hunk 1]
```diff
@@ -329,6 +215,129 @@ def test_no_interest(self):
...
+    @pytest.mark.parametrize('when, desired', [
+        (
+            None,
+            [-75.62318601, -76.25337923, -76.88882405, -77.52956425],
+        ), (
+            [0, 1, 'end', 'begin'],
+            [-75.62318601, -75.62318601, -76.88882405, -76.88882405],
+        )
+    ])
+    def test_broadcast(self, when, desired):
```
Comment (Kai-Striega) [inline numpy_financial/tests/test_financial.py]: This looks like it's testing only the first slice of the expected array (i.e. period 1) not each period for each when. I realise that the desired values are from the original tests, and that this is a refactoring, but I would prefer not to propagate an incorrect test.

i.e. I would expect the following two results to be the same, but they are not.

```python
In [7]: whens = [0, 1, 'begin', 'end']                                                                                                                                                                                  

In [8]: npf.ppmt(*args, whens)                                                                                                                                                                                          
Out[8]: array([-75.62318601, -75.62318601, -76.25337923, -77.52956425])

In [9]: for when in whens: 
   ...:     print(npf.ppmt(*args, when))                                                                                                                                                                                                            
[-75.62318601 -76.25337923 -76.88882405 -77.52956425]
[-91.52712662 -75.62318601 -76.25337923 -76.88882405]
[-91.52712662 -75.62318601 -76.25337923 -76.88882405]
[-75.62318601 -76.25337923 -76.88882405 -77.52956425]
```

Am I missing something obvious here, or are the existing tests incorrect?

Comment (person142) [reply numpy_financial/tests/test_financial.py]: Oh, IIUC you would only get the second result if you did something like `when = [[0], [1], ['end'], ['begin']]` or `numpy.arange(1, 5).reshape(4, 1)`. For example:

```python
>>> import numpy
>>> import numpy_financial as npf
>>> whens = [0, 1, 'end', 'begin']
>>> npf.ppmt(0.1 / 12, numpy.arange(1, 5).reshape(4, 1), 24, 2000, 0, whens)
array([[-75.62318601, -91.52712662, -75.62318601, -91.52712662],
       [-76.25337923, -75.62318601, -76.25337923, -75.62318601],
       [-76.88882405, -76.25337923, -76.88882405, -76.25337923],
       [-77.52956425, -76.88882405, -77.52956425, -76.88882405]])
```

That *is* a much better broadcasting test than the one that exists now.

File: numpy_financial/tests/test_financial.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
-
 from decimal import Decimal
+import math
```
[the remark this answers is not in the document]
Comment (person142) [reply numpy_financial/tests/test_financial.py]: Yup!

File: numpy_financial/tests/test_financial.py (hunk 1 above)
Comment (Kai-Striega) [reply numpy_financial/tests/test_financial.py]: That is what I had in mind, in many cases that would also much more useful. This also raises the question of what happens if more variables are included e.g. we should expect a 2 x 4 x4 array if we use rate = [0.1 / 12, 0.2 / 12].

Comment (Kai-Striega) [reply numpy_financial/tests/test_financial.py]: As this PR is for cleaning up - not changing the existing behaviour - I'm happy to mark this as resolved.

**Comment by Kai-Striega** [comment]
@person142 I'm happy with the changes as they are. Would you like to add any extra changes or are you happy for me to merge? Also, I lost track of this while working on something else (sorry!), if that happens in the future feel free to remind me and I'll try to move it up on my list of priorities.

**Comment by person142** [comment]
> Would you like to add any extra changes or are you happy for me to merge?

Let me improve that broadcasting test over the weekend. I’ve been meaning to do it but have been short on time.

**Comment by Kai-Striega** [comment]
@person142 I'm going to merge these changes as they are and open an issue for the broadcasting issue. Thanks for taking the time to neaten the tests so far, I do think it's heading in the right direction.

**Pull request merged**